### ISSUE TITLE
"delete" a user

### ISSUE THREAD

**Issue by pdpinch** [mitodl/micromasters#2018]
A user has requested that we delete his account. 

Having caused trouble with this before, what's the best way to do this? Does unsetting the user's `Active` bit do anything useful?

See https://odl.zendesk.com/agent/tickets/4336

**Comment by noisecapella** [comment]
I think that would prevent the user from logging in, though it won't delete the data attached.

I think the best solution for right now is to disable the rules protecting the audit tables, then delete the `User` and let that cascade. Since this user hasn't made a purchase we should not be deleting anything important.

Then we should change the audit table to store the user id instead of a foreign key itself, then reenable the protection of the audit tables.

**Comment by pdpinch** [comment]
I think preventing the user from logging in would be sufficient for these cases, for now. 

I'm going to do some tests with the active bit, to see what happens when it's turned on and off. 

Questions: 

- [ ] What happens when the user logs in with the active bit off?
- [ ] Can we turn it back on if/when the user changers their mind?
- [ ] Do the celery jobs still run that update the user's grades?

**Comment by pdpinch** [comment]
Some answers from testing:

1. With the active bit off, the user can't log in. Authenticating with edx.org succeeds, but after the user is returned to the portal, they are not logged in. 
2. Turning the active bit back on allows the user to login again without issue. 
3. Changing the active bit has no effect on the elastic search index. We should only index "active" users
4. I'm not sure how to test the celery job, but I will note that changing the active bit has no effect on the user's social auth credentials (unsurprisingly). It may be easiest to just check that the user is active before doing updates.

**Comment by pdpinch** [comment]
If we don't delete users outright, we'll need to update the TOS.

**Comment by giocalitri** [comment]
A possible approach could be to switch `User` model to a proxy model with a custom manager.
https://docs.djangoproject.com/en/1.11/topics/db/models/#proxy-models
https://docs.djangoproject.com/en/1.10/topics/auth/customizing/#reusable-apps-and-auth-user-model

Unfortunately changing the user model at this point will be quite difficult.
https://docs.djangoproject.com/en/1.10/topics/auth/customizing/#changing-to-a-custom-user-model-mid-project

but mostly this:
https://code.djangoproject.com/ticket/25313

**Comment by pdpinch** [comment]
We've moved discussion of this to https://docs.google.com/document/d/1Jyw-8FaHj6Cg2XSH6GteabjALRK1dsviTP-8k4bKIB0

I expect I'll open a new issue once the discussion is complete.

**Comment by pdpinch** [comment]
In #4139 we added a management command for retiring users.
